### PR TITLE
US89390 Make client lib more tolerant to network/landlord glitches

### PR DESCRIPTION
**Pull request by IhabAlkabary**
- [ ] This branch MUST NOT be merged until Landlord Prod healthcheck shows: 1.4.*
https://landlord.brightspace.com/health

### CODE REVIEW

**Comment by omsmith** [comment]
Trying to understand the reasoning behind the change. Healthchecks are red sometimes because of network blips?

**Comment by IhabAlkabary** [comment]
@omsmith We are seeing intermittent failures in this ping and no clear idea on why, so until we form a clear understanding, we thought we can make the the process more tolerant by adding timeout and retry and continue monitoring it.

**Comment by omsmith** [comment]
Given this is going to get used for an external service, isn't that just going to hide the fact that "at this particular time that you asked, I am actually unable to contact landlord"?

Is it just tcp pings that are having intermittent failures? Can landlord add an http route of "/ping" that we could hit instead that might be more stable?

**Comment by IhabAlkabary** [comment]
We already have the healthcheck for that, we are developing a healthcheck with "details" if you add a query param.

**Comment by IhabAlkabary** [comment]
Yes, I personally do not see issues with the healthcheck, just the ping

**Comment by omsmith** [comment]
So if I go to "https://landlord.brightspace.com/healthcheck" or whatever, it just returns 200 OK and does no other checks?

**Comment by omsmith** [comment]
This `validateConfiguration` function is used in many services in _their_ healthchecks, determining whether landlord can be reached.

I feel like adding retries hides that. If there's another mechanism for reaching out to landlord to verify connect-ability without causing any real processing on Landlord's side, which is different than a tcp ping, happy to switch to that.

**Comment by IhabAlkabary** [comment]
Yes, currently all it does is this:
https://landlord.brightspace.com/health

`{
application: {
identification: {
name: "Landlord",
version: "1.1.211"
},
health: {
general: {
status: "good",
description: ""
},
services: [
{
name: "Landlord Web",
description: "web service",
components: [ ],
metrics: [ ]
}
]
},
test: true
}
}`

**Comment by IhabAlkabary** [comment]
I would assume switching to pinging using the current health check will add a bit of load on Landlord. Wouldn't just pinging be lighter?

**Comment by omsmith** [comment]
Seems like the `/health` route:
1. Reads a file from disk
2. Makes a DB call to lookup my non-token

That's why I'm not suggesting calling landlords healthcheck.

This is an example of a `/ping` route we use in our services, which calling should caught approximately zero load.

```
const pingGet = route.unauthed.get(routes.ping.href,
	/* @this */ function *handlePingGet() {
		this.set('Cache-Control', 'no-cache');
		this.status = 200;
	}
);
```

I'm also happy to stick with the tcp-ping, but you're saying it fails sometimes and you have no indication why because as far as you can tell it shouldn't. I don't think retrying the tcp-ping three times is the correct direction to go in for a "please verify you can successfully contact your configured landlord location" function.

File: src/index.js [hunk 1]
```diff
@@ -158,24 +157,19 @@ LandlordClient.prototype.lookupTenantUrl = function lookupTenantUri(tenantId) {
...
-				reject(new errors.LandlordNotAvailable(self._landlord, data.results));
-			} else {
-				err ? reject(err) : resolve(data);
-			}
-		});
+		request
+			.get(self._landlord + '/ping')
+			.end(function(err, res) {
+				if (err) {
+					reject(new errors.LandlordNotAvailable(self._landlord, res));
```
Comment (omsmith) [inline src/index.js]: The `results` property of this is going to change, which is fine, but will make the change breaking.

File: test/index.spec.js [hunk 2]
```diff
@@ -138,4 +138,14 @@ describe('LandlordClient', function() {
 				expect(cacheTenantUrlLookupSpy.notCalled).to.be.true;
 			});
 	});
+
+	it('(ping) Landlord PROD is available', function() {
+		const instance = new LandlordClient({ endpoint: 'https://landlord.brightspace.com' });
```
Comment (omsmith) [inline test/index.spec.js]: I really would prefer the tests for the library not call out.

File: src/index.js (hunk 1 above)
Comment (IhabAlkabary) [inline src/index.js]: I can see if I can return the same structure to what the tcp ping was returning

File: src/index.js [hunk 3]
```diff
@@ -158,24 +157,19 @@ LandlordClient.prototype.lookupTenantUrl = function lookupTenantUri(tenantId) {
 		});
 };
 
-LandlordClient.prototype.validateConfiguration = function validateConfiguration(opts) {
+LandlordClient.prototype.validateConfiguration = function validateConfiguration() {
 	var self = this;
 	return new Promise(function(resolve, reject) {
-		const port = self._landlordOpts.port || (self._landlordOpts.protocol === 'https:' ? 443 : 80);
```
Comment (IhabAlkabary) [inline src/index.js]: done

File: src/index.js (hunk 1 above)
Comment (IhabAlkabary) [inline src/index.js]: who would be using the results returned from tcp -ping?
https://www.npmjs.com/package/tcp-ping

Comment (omsmith) [inline src/index.js]: Don't see anybody using it, and it's only on the error, but still a breaking change none-the-less.

**Comment by omsmith** [comment]
Landed as 96c9aa1..3fa7a72

**Pull request closed without merging**